test/jdk/java/lang/StableValue/StableFunctionTest.java

```diff
@@ -0,0 +1,219 @@
+        assertEquals(1, cif.cnt());
+        assertEquals(mapper.apply(Value.FORTY_TWO), cached.apply(Value.FORTY_TWO));
+        assertEquals(1, cif.cnt());
+        assertTrue(cached.toString().startsWith("{"), cached.toString());
+        // Key order is unspecified
+        assertTrue(cached.toString().contains(Value.THIRTEEN + "=.unset"));
+        assertTrue(cached.toString().contains(Value.FORTY_TWO + "=" + mapper.apply(Value.FORTY_TWO)), cached.toString());
+        assertTrue(cached.toString().endsWith("}"));
+        // One between the values
+        assertEquals(1L, cached.toString().chars().filter(ch -> ch == ',').count(), cached.toString());
+        var x0 = assertThrows(IllegalArgumentException.class, () -> cached.apply(Value.ILLEGAL_BEFORE));
+        assertTrue(x0.getMessage().contains("ILLEGAL"));
+        var x1 = assertThrows(IllegalArgumentException.class, () -> cached.apply(Value.ILLEGAL_BETWEEN));
+        assertTrue(x1.getMessage().contains("ILLEGAL"));
+        var x2 = assertThrows(IllegalArgumentException.class, () -> cached.apply(Value.ILLEGAL_AFTER));
+        assertTrue(x2.getMessage().contains("ILLEGAL"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("emptySets")
+    void empty(Set<Value> inputs) {
+        Function<Value, Integer> f0 = StableValue.function(inputs, Value::asInt);
+        Function<Value, Integer> f1 = StableValue.function(inputs, Value::asInt);
+        assertTrue(f0.toString().contains("{}"));
+        assertThrows(NullPointerException.class, () -> f0.apply(null));
+        assertNotEquals(f0, f1);
+        assertNotEquals(null, f0);
+    }
+
+    @ParameterizedTest
+    @MethodSource("nonEmptySets")
+    void exception(Set<Value> inputs) {
+        StableTestUtil.CountingFunction<Value, Integer> cif = new StableTestUtil.CountingFunction<>(_ -> {
+            throw new UnsupportedOperationException();
+        });
+        var cached = StableValue.function(inputs, cif);
+        assertThrows(UnsupportedOperationException.class, () -> cached.apply(Value.FORTY_TWO));
+        assertEquals(1, cif.cnt());
+        assertThrows(UnsupportedOperationException.class, () -> cached.apply(Value.FORTY_TWO));
+        assertEquals(2, cif.cnt());
+        assertTrue(cached.toString().startsWith("{"));
+        // Key order is unspecified
+        assertTrue(cached.toString().contains(Value.THIRTEEN + "=.unset"));
+        assertTrue(cached.toString().contains(Value.FORTY_TWO + "=.unset"), cached.toString());
+        assertTrue(cached.toString().endsWith("}"));
+    }
+
+    @ParameterizedTest
+    @MethodSource("nonEmptySets")
+    void circular(Set<Value> inputs) {
+        final AtomicReference<Function<?, ?>> ref = new AtomicReference<>();
+        Function<Value, Function<?, ?>> cached = StableValue.function(inputs, _ -> ref.get());
+        ref.set(cached);
+        cached.apply(Value.FORTY_TWO);
+        String toString = cached.toString();
```

Comment:
I'd recommend caching the toString() representation in the other test cases as well :+1: